web/packages/teleterm/src/ui/DocumentCluster/useUserPreferences.ts

```diff
@@ -0,0 +1,189 @@
+
+import { UserPreferences } from 'teleterm/services/tshd/types';
+
+export function useUserPreferences(clusterUri: ClusterUri): {
+  userPreferencesAttempt: Attempt<UserPreferences>;
+  unifiedResourcePreferencesFallback: UnifiedResourcePreferences;
+  updateUserPreferencesAttempt: Attempt<void>;
+  updateUserPreferences(newPreferences: UserPreferences): Promise<void>;
+} {
+  const appContext = useAppContext();
+  const [
+    fetchUserPreferencesAttempt,
+    runFetchUserPreferencesAttempt,
+    setFetchUserPreferencesAttempt,
+  ] = useAsync<Promise<UserPreferences>[], UserPreferences>(value => value);
+  const [updateUserPreferencesAttempt, runUpdateUserPreferencesAttempt] =
+    useAsync(async (newPreferences: UserPreferences) =>
+      appContext.tshd.updateUserPreferences({
+        clusterUri,
+        userPreferences: newPreferences,
+      })
+    );
+  appContext.workspacesService.useState();
+
+  const getPreferencesPromise = useRef<ReturnType<typeof getPreferences>>();
+  const getPreferences = useCallback(async () => {
+    const preferencesPromise = appContext.tshd.getUserPreferences({
+      clusterUri,
+    });
+    getPreferencesPromise.current = preferencesPromise;
+    return preferencesPromise;
+  }, [appContext.tshd, clusterUri]);
+
+  const unifiedResourcePreferencesFallback =
+    appContext.workspacesService.getUnifiedResourcePreferences(
+      routing.ensureRootClusterUri(clusterUri)
+    ) || {
+      defaultTab: DefaultTab.DEFAULT_TAB_ALL,
+      viewMode: ViewMode.VIEW_MODE_CARD,
+    };
+
+  const updateFetchAttemptAndWorkspace = useCallback(
+    async (preferencesPromise: Promise<UserPreferences>) => {
+      const [updated, error] = await runFetchUserPreferencesAttempt(
+        preferencesPromise
+      );
+      if (!error) {
+        appContext.workspacesService.setUnifiedResourcePreferences(
+          routing.ensureRootClusterUri(clusterUri),
+          updated.unifiedResourcePreferences
+        );
+      }
+    },
+    [appContext.workspacesService, clusterUri, runFetchUserPreferencesAttempt]
+  );
+
+  useEffect(() => {
+    if (fetchUserPreferencesAttempt.status === '') {
+      updateFetchAttemptAndWorkspace(getPreferences());
+    }
+  }, [
+    updateFetchAttemptAndWorkspace,
+    fetchUserPreferencesAttempt.status,
+    getPreferences,
+  ]);
+
+  async function updateUserPreferences(
+    newPreferences: Partial<UserPreferences>
+  ): Promise<void> {
+    // If we don't have the preferences fetched yet,
+    // we start from updating the workspace (so the user sees an update in the UI).
+    //
+    // Since we know that the preferences have changed,
+    // we don't want to show the user the preferences from the "get"
+    // request that are outdated.
+    // Instead, we replace a request in that attempt with an "update" request,
+    // falling back to "get" response, only if the update failed.
+    // Thanks to that, the user sees a continuous loading screen.
```

Comment:
The code which achieves that is a bit hard to understand, mostly because it uses some approaches which are not common in our codebase (`useAsync((value) => value)`, storing a promise in a ref).

The way I understand this problem is that if someone clicks a button to change from cards to a list but we're yet to know what the actual preferences are like, we want to disregard the actual prefs we get from the cluster and wait for the updated prefs instead.

Isn't this a matter of having two different attempts, one that fetches the current prefs and another that updates them, and returning only one of them from the hook? If an update is in progress, we take the prefs from the update response. If there's no update, we use the actual prefs.